provider/kubernetes/client.go

```diff
@@ -150,31 +125,37 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
 		c.isNamespaceAll = true
 	}
 
-	var informManager informerManager
+	var wg sync.WaitGroup
+	eventHandler := newResourceEventHandler(eventCh)
 	for _, ns := range namespaces {
-		ns := ns
-		informManager.extend(c.WatchIngresses(ns, kubeLabelSelector, eventCh), true)
-		informManager.extend(c.WatchObjects(ns, kindServices, &corev1.Service{}, c.svcStores, eventCh), true)
-		informManager.extend(c.WatchObjects(ns, kindEndpoints, &corev1.Endpoints{}, c.epStores, eventCh), true)
-		// Do not wait for the Secrets store to get synced since we cannot rely on
-		// users having granted RBAC permissions for this object.
-		// https://github.com/containous/traefik/issues/1784 should improve the
-		// situation here in the future.
-		informManager.extend(c.WatchObjects(ns, kindSecrets, &corev1.Secret{}, c.secStores, eventCh), false)
+		factory := informers.NewFilteredSharedInformerFactory(c.clientset, resyncPeriod, ns, func(opts *metav1.ListOptions) {
+			opts.LabelSelector = labelSelector
+		})
+		factory.Extensions().V1beta1().Ingresses().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Services().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Endpoints().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Secrets().Informer().AddEventHandler(eventHandler)
+		c.factories[ns] = factory
 	}
 
-	var wg sync.WaitGroup
-	for _, informer := range informManager.informers {
-		informer := informer
+	for _, ns := range namespaces {
 		safe.Go(func() {
 			wg.Add(1)
-			informer.Run(stopCh)
+			c.factories[ns].Start(stopCh)
```

Comment:
We need to insert `ns := ns` after line 141 in order to capture the variable. Otherwise, we'd always end up with the last namespace being referenced in the anonymous goroutine (see also https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables).

Reply:
I just realise that `informerFactory.Start` is an asynchronous call itself and we don't have to invoke it via `safe.Go(...)`